crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs

```diff
@@ -156,6 +205,10 @@ impl Context {
     /// parameters. This is because block parameters are converted before anything else, and
     /// because instructions results are converted when the corresponding instruction is
     /// encountered. (An instruction result cannot be referenced before the instruction occurs.)
+    ///
+    /// It is not safe to call this function on value ids that represent addresses. Instructions
+    /// involving such values are evaluated via a separate path and stored in
+    /// `ssa_value_to_array_address` instead.
```

Comment:
The complication is that a `ValueId` could either map to an `AcirVar` (i.e. a constant, witness or expression), or it could map to an array address (comprised of an ArrayId plus index offset).

I don't think there's are pretty way of rationalising this relation. If I were to force the array address into an `AcirVar`, I'd be pushing the confusion one layer deeper, and `AcirContext` would have perform many checks for its various operations. At least in `acir_gen/mod` we have context on whether some value relates to an array address or not.

I could combine `ssa_value_to_acir_var` and `ssa_value_to_array_address` into a single map via an enum to distinguish between addresses and acir vars. That would at least force more safety checks... Thoughts?

Reply:
> I could combine ssa_value_to_acir_var and ssa_value_to_array_address into a single map via an enum to distinguish between addresses and acir vars

Could be beneficial, not sure. I like the idea of more safety checks but I'll leave it to you as to whether you think it is an improvement or not.

Reply:
I'll keep as is for now. I dare say dynamic arrays are going to force us to reconsider these relationships again anyway - I'd rather wait until we have more context.